modules/globebrowsing/util/globelabelscomponent.cpp

```diff
@@ -0,0 +1,846 @@
+            );
+            fadeRange.x += _labelsFadeOutDist;
+            double a = rangeAngularCoefConst / (fadeRange.x - fadeRange.y);
+            double b = -(fadeRange.y / (fadeRange.x - fadeRange.y));
+            double funcValue = a * distanceCameraGlobeWorld + b;
+            varyingOpacity *= funcValue > 1.0 ? 1.f : static_cast<float>(funcValue);
+
+            if (varyingOpacity < minTransparencyValueConst) {
+                return;
+            }
+        }
+
+        renderLabels(
+            data, 
+            mvp, 
+            static_cast<float>(distanceCameraGlobeWorld), 
+            varyingOpacity
+        );
+    }
+}
+
+void GlobeLabelsComponent::renderLabels(const RenderData& data,
+                                        const glm::dmat4& modelViewProjectionMatrix, 
+                                        float distToCamera,
+                                        float fadeInVariable
+) {
+
+    glm::vec4 textColor = _labelsColor;
+    textColor.a *= fadeInVariable;
+    constexpr double DIST_EPS = 6000.0;
+    constexpr double SIN_EPS = 0.001;
+
+    glm::dmat4 invMP = glm::inverse(_globe->modelTransform());
+    glm::dmat4 invCombinedView = glm::inverse(data.camera.combinedViewMatrix());
+
+    glm::dvec4 cameraPosWorld = invCombinedView * glm::dvec4(0.0, 0.0, 0.0, 1.0);
+    glm::dvec3 cameraPosObj = glm::dvec3(invMP * cameraPosWorld);
+    glm::dvec4 cameraUpVecWorld = glm::dvec4(data.camera.lookUpVectorWorldSpace(), 0.0); 
+    glm::dvec3 cameraLookUpObj = glm::dvec3(invMP * cameraUpVecWorld);
+
+    glm::dmat4 VP = glm::dmat4(
+        data.camera.sgctInternal.projectionMatrix()
+    ) * data.camera.combinedViewMatrix();
+
+
+    glm::dmat4 invModelMatrix = glm::inverse(_globe->modelTransform());
+
+    glm::dvec3 cameraViewDirectionObj = glm::dvec3(
+        invModelMatrix * glm::dvec4(data.camera.viewDirectionWorldSpace(), 0.0)
+    );
+    glm::dvec3 cameraUpDirectionObj = glm::dvec3(
+        invModelMatrix * glm::dvec4(data.camera.lookUpVectorWorldSpace(), 0.0)
+    );
+    glm::dvec3 orthoRight = glm::normalize(
+        glm::cross(cameraViewDirectionObj, cameraUpDirectionObj)
+    );
+    if (orthoRight == glm::dvec3(0.0)) {
+        glm::dvec3 otherVector(
+            cameraUpDirectionObj.y,
+            cameraUpDirectionObj.x,
+            cameraUpDirectionObj.z
+        );
+        orthoRight = glm::normalize(glm::cross(otherVector, cameraViewDirectionObj));
+    }
+    glm::dvec3 orthoUp = glm::normalize(
+        glm::cross(orthoRight, cameraViewDirectionObj)
+    );
+
+    for (const LabelEntry lEntry : _labels.labelsArray) {
+        glm::vec3 position = lEntry.geoPosition;
+        glm::dvec3 locationPositionWorld =
+            glm::dvec3(_globe->modelTransform() * glm::dvec4(position, 1.0));
+        double distanceCameraToLabelWorld =
+            glm::length(locationPositionWorld - data.camera.positionVec3());
+
+        bool draw = false;
```

Comment:
Better to split this into two variables and then || them in the if statement on 727